Support psycopg as a PostgreSQL driver

clokep opened the pull request:
This adds support for psycopg (aka psycopg3) as a PostgreSQL driver in addition to psycopg2. This requires adding some abstractions between psycopg2 and psycopg although most of the queries can be identical (since it is just PostgreSQL underneath).

This is a branch that @realtyem and myself have been working on for several years.

The bulk of this PR is splitting the `PostgresEngine` into a `PsycopgEngine` and `Psycopg2Engine` which use the correct underlying driver. We then need to update _some_ if-statements in the code to differentiate between these.

A bunch of configuration code also needs to be adjusted to handle having multiple engines for postgres.

This also updates the test matrix for trial to run against psycopg2 and psycopg.

This is step 1 of #14586.

clokep committed:
Add psycopg package.

clokep committed:
Separate PostgresEngine into Psycopg2Engine and PsycopgEngine.

clokep committed:
Run tests in CI against psycopg.

clokep committed:
Method to set statement timeout.

clokep committed:
Update user directory to handle psycopg3.

clokep committed:
Support execute_values on psycopg3.

clokep committed:
Fix-up simple_* tests.

clokep committed:
Fix-up calls to end_to_end_keys.

realtyem committed:
Use superclass version of `executescript()` (#1)

realtyem committed:
Switch out formatting placeholder for what psycopg2 is expecting (#2)

clokep committed:
Merge remote-tracking branch 'upstream/develop' into psycopg3

realtyem committed:
Linting (and a fix) (#3)

Use the proper isolation level on psycopg, where it is no longer an int.

clokep committed:
Merge remote-tracking branch 'refs/remotes/origin/psycopg3' into psycopg3

clokep committed:
Try running complement builds?

realtyem committed:
Merge branch 'develop' into psycopg3

realtyem committed:
Adjust type: ignore line to where mypy will apply it

realtyem committed:
Add both PsycopgEngine and Psycopg2Engine to database.engines.__all__ for importing

realtyem committed:
Adjust _mark_state_groups_as_pending_deletion_txn() to use execute_batch() instead of execute_values()

execute_values() from psycopg2 can do insertions with fetch=False, but
psycopg is better off inserting with execute_batch() instead(because of
pipelining support)

realtyem committed:
Adjust set_profile_field() insertion sql for a narrower type on parameter $3

realtyem committed:
Try an update to the config schema?

realtyem committed:
Adjust unit tests to reflect updated minimal versions of python/postgres

clokep committed:
Minor changes to profile field

clokep committed:
Merge remote-tracking branch 'upstream/develop' into psycopg3

clokep committed:
Newsfragment

clokep committed:
Merge remote-tracking branch 'upstream/develop' into psycopg3

clokep committed:
poetry lock again

clokep committed:
Merge branch 'develop' into psycopg3

reivilibre reviewed (changes requested):
Thanks both of you, looks like a tremendous amount of work has gone into this!

From my perspective, this looks virtually ready, just the really minor things I've pointed out, I'd say:

- as said, I would like more eyes on the `COPY`-based replacement for batch execution. It is probably fine, though.
- It looks like some optimisations have been walked back compared to Psycopg2. I personally think this is fine, though it would be ideal to have a tracking issue containing a ticklist of these, so we can track their re-implementation as a prerequisite to considering Psycopg support 'complete'/first-class/ready to replace Psycopg2.

Other than that, it all looks clear and reasonable to me!

reivilibre reviewed (changes requested):
Thanks both of you, looks like a tremendous amount of work has gone into this!

From my perspective, this looks virtually ready, just the really minor things I've pointed out, I'd say:

- as said, I would like more eyes on the `COPY`-based replacement for batch execution. It is probably fine, though.
- It looks like some optimisations have been walked back compared to Psycopg2. I personally think this is fine, though it would be ideal to have a tracking issue containing a ticklist of these, so we can track their re-implementation as a prerequisite to considering Psycopg support 'complete'/first-class/ready to replace Psycopg2.

Other than that, it all looks clear and reasonable to me!

realtyem committed:
Merge branch 'develop' into psycopg3-adjust-settings

realtyem committed:
Update poetry.lock file

realtyem committed:
linting fixups

clokep committed:
Merge pull request #4 from realtyem/psycopg3-adjust-settings

jason-famedly committed:
Merge branch 'develop' into jason/psycopg-merge-develop

jason-famedly committed:
Adjust for execute_values() usage with fetch=False to only be appropriate on psycop2 for now

This will use executemany()(inside of execute_batch()) from psycopg
instead, which uses pipeline mode

jason-famedly committed:
Attempt forcing a lower case comparison to "psycopg" in complement.sh

jason-famedly committed:
Add mini docstrings to the abstract postgres engine class, as well as the concrete definitions

jason-famedly commented on synapse/storage/engines/postgres.py:
```diff
@@ -75,6 +75,10 @@ def _disable_bytes_adapter(_: bytes) -> NoReturn:
 
+    def set_statement_timeout(self, cursor: Cursor, statement_timeout: int) -> None:
+        """Configure the current cursor's statement timeout."""
+        cursor.execute("SET statement_timeout TO ?", (statement_timeout,))
```
When the time comes to remove psycopg2, there is a property that will be set instead. I do not recall why we did not do this here, but probably can rectify in future work. It's probable that at the time of initial bring up, that property did not exist yet(but I would not quote that)

MadLittleMods replied:
This point still seems relevant.

But the prior art uses `SET statement_timeout` so I don't think this is a blocker

jason-famedly commented on scripts-dev/complement.sh:
```diff
@@ -256,7 +260,9 @@ if [[ -n "$WORKERS" ]]; then
 else
   export PASS_SYNAPSE_COMPLEMENT_USE_WORKERS=
-  if [[ -n "$POSTGRES" ]]; then
+  if [[ "$POSTGRES" = "psycopg" ]]; then
```
I imagine you are correct. I think I was using it from the command line and tested with it's lower case function. I will adjust to use the bash `"${POSTGRES,,}}"` pattern instead, which should lower case all the letters

MadLittleMods replied:
Resolving as:

Case-insensitive conditions were added :white_check_mark: 

Comment was added with this context :white_check_mark:

jason-famedly committed:
I forgot that the new driver is not wired up in the complement script for use with workers yet

clokep committed:
Merge pull request #6 from jason-famedly/jason/psycopg-merge-develop

jason-famedly committed:
Use the 'c' extra instead of the 'pure python' version for C-compiled speedups

jason-famedly committed:
Update 'c' extra to 3.2.8 as a minimum. This avoids olddebian's packaged libpq5 having Postgres 18 support which breaks previous versions of psycopg when compiled

jason-famedly committed:
Allow complement to be ran with WORKERS=1 and still allow the psycopg driver(in POSTGRES=psycopg)

clokep committed:
Merge pull request #7 from jason-famedly/jason/adjust-for-feedback

jason-famedly committed:
Merge branch 'develop' into jason/update-psycopg3

jason-famedly committed:
update poetry lock

jason-famedly committed:
Merge branch 'develop' into jason/update-psycopg3

clokep committed:
Merge pull request #8 from jason-famedly/jason/update-psycopg3

jason-famedly committed:
Merge branch 'develop' into jason/update-poetry-lock

clokep committed:
Merge pull request #9 from jason-famedly/jason/update-poetry-lock

MadLittleMods commented on docker/conf/homeserver.yaml:
```diff
@@ -66,11 +66,11 @@ listeners:
     user: "{{ POSTGRES_USER or "synapse" }}"
     password: "{{ POSTGRES_PASSWORD }}"
-    database: "{{ POSTGRES_DB or "synapse" }}"
+    dbname: "{{ POSTGRES_DB or "synapse" }}"
```
I assume `dbname` is a cross-compatible standard arg vs `database`?

jason-famedly replied:
More or less, yes. `database` was never supposed to be a supported parameter for Postgres based drivers that connect through `libpq`. It worked, but it was not supposed to. All these values are passed into `libpq` to generate the connection. I assume(but have not dug far in looking) that this is something handled internally by the DBAPI2.0 compatible module. `psycopg` rejects this value, where `psycopg2` did not. This change in particular was just to make sure the parameters were unified to what the changes in sytest became.

jason-famedly replied:
Further context: https://github.com/matrix-org/synapse/pull/16618

MadLittleMods commented on changelog.d/18999.feature:
```diff
@@ -0,0 +1 @@
+Support using psycopg as the PostgreSQL driver. Contributed by @clokep and @realtyem.
```
> It looks like some optimisations have been walked back compared to Psycopg2. I personally think this is fine, though it would be ideal to have a tracking issue containing a ticklist of these, so we can track their re-implementation as a prerequisite to considering Psycopg support 'complete'/first-class/ready to replace Psycopg2.
>
> *-- @reivilibre, https://github.com/element-hq/synapse/pull/18999#pullrequestreview-3314480569*

To re-explain for myself, we've updated a bunch of spots from `if isinstance(self.database_engine, PostgresEngine):`  to `if isinstance(self.database_engine, Psycopg2Engine):` without any optimized alternative for `psycopg` (just uses the slow fallback at the moment).

TODO: ^ make list, make issue, etc

MadLittleMods commented on synapse/storage/engines/postgres.py:
```diff
@@ -40,19 +41,16 @@
+    Generic[ConnectionType, CursorType, IsolationLevelType],
+    BaseDatabaseEngine[ConnectionType, CursorType, IsolationLevelType],
+    metaclass=abc.ABCMeta,
 ):
```
Docstring added :white_check_mark: 

https://github.com/clokep/synapse/blob/59cd733fc95c4dc69fd7f5b8d20cdbc05f938518/synapse/storage/engines/postgres.py#L44-L53

MadLittleMods commented on synapse/storage/databases/main/end_to_end_keys.py:
```diff
@@ -1303,7 +1303,7 @@ def _claim_e2e_fallback_keys_bulk_txn(
             SET used = used OR mark_as_used
             FROM claims
             WHERE (k.user_id, k.device_id, k.algorithm) = (claims.user_id, claims.device_id, claims.algorithm)
-            RETURNING k.user_id, k.device_id, k.algorithm, k.key_id, k.key_json;
```
Asking for reference: Does this sort of change actually matter? (cause syntax error problems with the new driver?)

Or just a clean-up

jason-famedly replied:
Cleanup, if memory serves. There was a brief time where `execute_batch()` was used instead of the newer `COPY` based methods for `execute_values()`, there may have been some kind of issue with two `;` appending two separate statements that caused an issue. But I do not recall exactly.

MadLittleMods commented on synapse/storage/databases/main/profile.py:
```diff
@@ -455,12 +460,23 @@ def set_profile_field(txn: LoggingTransaction) -> None:
                 # keys will be taken from the second value.
                 sql = """
-                INSERT INTO profiles (user_id, full_user_id, fields) VALUES (?, ?, JSON_BUILD_OBJECT(?, ?::jsonb))
+                INSERT INTO profiles (user_id, full_user_id, fields) VALUES (?, ?, JSON_BUILD_OBJECT(?::text, ?::jsonb))
```
Does this change matter or is this just being more explicit?

jason-famedly replied:
`psycopg.errors.IndeterminateDatatype: could not determine data type of parameter $3` suggested that the hint was important. `psycopg2` did not care or coerces it in a different way perhaps

MadLittleMods commented on synapse/storage/databases/main/user_directory.py:
```diff
@@ -693,37 +698,69 @@ def _update_profiles_in_user_dir_txn(
-            txn.execute_values(
-                sql,
-                [
+            if isinstance(self.database_engine, Psycopg2Engine):
```
Why do we have this differentiation? (comment)

jason-famedly replied:
Browsing over this again, I believe this may have been written before there was a proper `execute_values()` replacement and before the helper to replace `%s` with `?` was prepared 🤔 It may be that this can be revisited. I will investigate

jason-famedly replied:
Oh, I see why now. This is an `INSERT` which is not supported by `execute_values()` as written for `psycopg`. Hence, the `fetch` parameter would have needed to be `False`. `executemany()` should run in it's pipelined mode which should be reasonably fast. I'll see about summarizing that into a comment

jason-famedly replied:
Added in [0de805a](https://github.com/element-hq/synapse/pull/18999/commits/0de805ab9ad59e4bc1672d8d8aa3ec5dd4f826fb)

MadLittleMods commented on synapse/storage/engines/__init__.py:
```diff
@@ -56,7 +68,10 @@ def create_engine(database_config: Mapping[str, Any]) -> BaseDatabaseEngine:
-        return PostgresEngine(database_config)
+        return Psycopg2Engine(database_config)
+
+    if name == "psycopg":
```
For sanity for downstream consumers of Synapse (in their configs), why not use the alias `psycopg3`?

jason-famedly replied:
Largely, because there is no actual such thing as `psycopg3`. That `psycopg2` existed with a numeral on the end was the [mistake](https://github.com/psycopg/psycopg2/issues/1632). As this is the exact string that is used to reference the module, and to assist in further reducing exposure to something that does not exist, I endeavor to prefer using the one with no numeral

This subject may be touched on again in the future though, per the conversation [in the RFC issue](https://github.com/element-hq/synapse/issues/14586#issuecomment-3381404080)

MadLittleMods commented on synapse/storage/engines/_base.py:
```diff
@@ -123,7 +126,7 @@ def attempt_to_set_autocommit(self, conn: ConnectionType, autocommit: bool) -> N
     @abc.abstractmethod
     def attempt_to_set_isolation_level(
-        self, conn: ConnectionType, isolation_level: int | None
+        self, conn: ConnectionType, isolation_level: IsolationLevel | None = None
```
Doesn't seem like we should have a default

MadLittleMods replied:
Still relevant

jason-famedly replied:
Having a default allowed for the abstraction at the postgres level to be used, instead of breaking both the `psycopg2` and `psycopg` drivers to have their own in `on_new_connection()`. It was an effort to avoid code duplication, in other words, and modeled(somewhat) pre-existing behavior. If you think it should be broken out into each respective drivers model, then that can be done.

MadLittleMods commented on synapse/storage/engines/psycopg.py:
```diff
@@ -0,0 +1,105 @@
+):
+    """
+    The class that defines how Synapse will interact with Postgres using the psycopg
+    version 3+ module.
```
For easy `Ctrl+f` finding,

```suggestion
    version 3+ module (aka psycopg3).
```

MadLittleMods replied:
Resolving

See https://github.com/element-hq/synapse/pull/18999#discussion_r3314136258 for why no `psycopg3`

MadLittleMods commented on synapse/storage/engines/psycopg.py:
```diff
@@ -0,0 +1,105 @@
+
+    def convert_param_style(self, sql: str) -> str:
+        # if isinstance(sql, psycopg.sql.Composed):
+        #     return sql
```
Leftover code relevant?

jason-famedly replied:
It should not be. `psycopg` has stricter requirements for `LiteralString` than `psycopg2`, and has compose-able SQL helpers to assist in avoid injection attacks. It turned out this was not needed here, but something similar somewhere *may* still be desired in the future

jason-famedly replied:
Removed in [0a3cd96](https://github.com/element-hq/synapse/pull/18999/commits/0a3cd96996054d6d88c1faa49108cd266f99ad57)

MadLittleMods commented on synapse/storage/engines/psycopg.py:
```diff
@@ -0,0 +1,105 @@
+        self, conn: psycopg.Connection, autocommit: bool
+    ) -> None:
+        # Sometimes this gets called with a Twisted connection instead, unwrap
+        # it because it doesn't support __setattr__.
```
Is this our own doing? Why isn't the linting catching this?

Can we instead fix up the spots in Synapse where this happens?

jason-famedly replied:
This is complicated. And expected. All connections that pass through all database engine definitions of Synapse are actually Twisted Database Pool Connections. The Twisted Connection and the `LoggingDatabaseConnection` has the proxying `__getattr__()` that pulls things out from the connection being wrapped but no companion `__setattr__()`, so *methods* work but *property attributes* seem not to. That is the 'quick and dirty' explanation.

That said, the easy solution is to switch to the `set_autocommit()` function here instead of accessing the attribute directly. This function was introduced in `psycopg` 3.2 which did not exist at the time this original code was written up. It allows all these conditionals to go away.

My to-do list for followups has Typing work for all of these engines on it. But what that should end up looking like and if it is helpful or not I could not say.

jason-famedly replied:
I will add that there is one known exception to this: during start of the server when the `DataStore` is being constructed. The pool does not exist yet at this time, so it will obviously not be a pool connection then

jason-famedly replied:
Adjusted to use the method setter instead of attribute in [6952653](https://github.com/element-hq/synapse/pull/18999/commits/69526539f5bd1a87793ee7e33fd95e54a48d38b9)

jason-famedly replied:
Had a follow up lint to remove the now unused import in [595bd8d](https://github.com/element-hq/synapse/pull/18999/commits/595bd8db7acdcc209e66504f7b8b798a48897b62)

MadLittleMods commented on synapse/storage/engines/_base.py:
```diff
@@ -123,7 +126,7 @@ def attempt_to_set_autocommit(self, conn: ConnectionType, autocommit: bool) -> N
+        self, conn: ConnectionType, isolation_level: IsolationLevel | None = None
     ) -> None:
         """Attempt to set the connections isolation level.
 
```
We should note that `None` means we return to the default isolation level

jason-famedly replied:
docstring updated in [0de805a](https://github.com/element-hq/synapse/pull/18999/commits/0de805ab9ad59e4bc1672d8d8aa3ec5dd4f826fb)

MadLittleMods commented on synapse/storage/background_updates.py:
```diff
@@ -788,6 +788,8 @@ def create_index_psql(conn: "LoggingDatabaseConnection") -> None:
             # postgres insists on autocommit for the index
             conn.engine.attempt_to_set_autocommit(conn.conn, True)
 
+            assert isinstance(self.db_pool.engine, PostgresEngine)
```
Possible to have a similar assertion in `create_index_sqlite(...)`?

jason-famedly replied:
Added in [e14c5a3](https://github.com/element-hq/synapse/pull/18999/commits/e14c5a30f0cc7bcec770b3a0f16a91233d8e0cac)

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -408,7 +416,7 @@ def execute_batch(self, sql: str, args: Iterable[Iterable[Any]]) -> None:
         """
 
-        if isinstance(self.database_engine, PostgresEngine):
+        if isinstance(self.database_engine, Psycopg2Engine):
```
Why specific to vs `Psycopg2Engine` vs the generic `PostgresEngine`? (comment)

jason-famedly replied:
Added a couple of comments in [1331918](https://github.com/element-hq/synapse/pull/18999/commits/1331918d11687e6054d9e82ee28db663d97f0e91)

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -442,17 +452,60 @@ def execute_values(
+                sql,
+                values,
+            )
+        else:
```
Feels like this should be an exact check and have a `assert_never(...)` for the `else` clause.

jason-famedly replied:
What do you mean? It is asserted just above that it must be postgres, and there are only the two options?

MadLittleMods replied:
Search for `assert_never(...)` in the codebase

It helps when the code yells at you when there are changes to what engines are supported like we're doing in this PR.

jason-famedly replied:
Changed in [7afece6](https://github.com/element-hq/synapse/pull/18999/commits/7afece6261250be5a3840ccb025031e7f595e01d)

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -442,17 +452,60 @@ def execute_values(
+                    yield from copy.rows()
+
+            # Flatten the values.
+            return self._do_execute(f, sql, list(itertools.chain.from_iterable(values)))
```
Why do we have all of this compared to the previous implementation?

I assume there isn't any `psycopg2.extras.execute_values` equivalent for `psycopg3` (there should be a comment about this)

MadLittleMods replied:
Resolving as I think this was answered by the comments added here :fast_forward:

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -442,17 +452,60 @@ def execute_values(
+            assert fetch is True
+
+            # execute_values requires a single replacement, but we need to expand it
+            # for COPY. This assumes all inner sequences are the same length.
```
> This assumes all inner sequences are the same length.

We should probably assert this property if we're relying on it

jason-famedly replied:
Added some quick assertions in [4305c48](https://github.com/element-hq/synapse/pull/18999/commits/4305c480cac5f098f26c166dd0f27fa334671420) and a comment about what part can be removed later

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -442,17 +452,60 @@ def execute_values(
+            value_str = "(" + ", ".join("?" for _ in next(iter(values))) + ")"
+            sql = sql.replace("?", ", ".join(value_str for _ in values))
+
+            # Wrap the SQL in the COPY statement.
+            sql = f"COPY ({sql}) TO STDOUT"
```
We should comment why we do this. I assume something about streaming so we don't have to construct a single giant list to send over

jason-famedly replied:
Not sure if I got the *why* quite right, but did take the opportunity of detailing a bit more of *how* it works in [b706c1d](https://github.com/element-hq/synapse/pull/18999/commits/b706c1d8e79963bae93c8b67190e3f564751777b)

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -410,17 +419,56 @@ def execute_values(
-            lambda the_sql, the_values: execute_values(
-                self.txn, the_sql, the_values, template=template, fetch=fetch
-            ),
-            sql,
-            values,
-        )
+        if isinstance(self.database_engine, Psycopg2Engine):
+            from psycopg2.extras import execute_values
+
+            return self._do_execute(
+                # TODO: is it safe for values to be Iterable[Iterable[Any]] here?
+                # https://www.psycopg.org/docs/extras.html?highlight=execute_batch#psycopg2.extras.execute_values says values should be Sequence[Sequence]
+                lambda the_sql, the_values: execute_values(
+                    self.txn, the_sql, the_values, template=template, fetch=fetch
+                ),
+                sql,
+                values,
+            )
+        else:
+            # We use fetch = False to mean a writable query. You *might* be able
+            # to morph that into a COPY (...) FROM STDIN, but it isn't worth the
+            # effort for the few places we set fetch = False.
+            assert fetch is True
+
+            # execute_values requires a single replacement, but we need to expand it
+            # for COPY. This assumes all inner sequences are the same length.
+            value_str = "(" + ", ".join("?" for _ in next(iter(values))) + ")"
+            sql = sql.replace("?", ", ".join(value_str for _ in values))
+
+            # Wrap the SQL in the COPY statement.
+            sql = f"COPY ({sql}) TO STDOUT"
+
+            def f(
+                the_sql: str, the_args: Sequence[Sequence[Any]]
+            ) -> Iterable[Tuple[Any, ...]]:
+                with self.txn.copy(the_sql, the_args) as copy:
+                    yield from copy.rows()
+
+            # Flatten the values.
+            return self._do_execute(f, sql, list(itertools.chain.from_iterable(values)))
```
Is this based on anything?

Why can't we use this new implementation for all Postgres drivers?

jason-famedly replied:
Unfortunately, `psycopg2` support for `COPY` is a little lacking. It could probably be coaxed into a usable form(and I have been experimenting with that with mixed results), but it would require using "filelike" objects(such as `io.StringIO`) to pass parameters around. This felt like a memory issue waiting to happen. [Psycopg2 docs on COPY](https://www.psycopg.org/docs/usage.html#using-copy-to-and-copy-from)

Consider that `execute_values()` allows passing a query in, and it will work as it looks like it should. `COPY` does not work that way for inserting data, so not only would the `VALUES` sequence need to be adjusted as it is now, but the verb `INSERT INTO` would have to be stripped off. Then there are upserts, I do not know how those behave yet and am not sure I trust that it would "just work". [Postgres docs on COPY](https://www.postgresql.org/docs/current/sql-copy.html). Inserting data would have to look like exporting data to wrap the query in `COPY`.

Also consider, for the moment, the `template` support here is incomplete. Again, for the moment, this is fine as no queries in Synapse use `fetch=True` with a `template`. It is coming in the future, but was not needed yet and the size of this work was large enough.

So, the potential exists that a `COPY` implementation for `psycopg2` would be doable, in theory. It needs a bit more research and testing to make it generic enough

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -442,17 +452,60 @@ def execute_values(
+    def copy_write(
+        self, sql: str, args: Iterable[Any], values: Iterable[Iterable[Any]]
+    ) -> None:
+        """Corresponds to a PostgreSQL COPY (...) FROM STDIN call."""
```
Comment why only available to to `PsycopgEngine` (`psycopg3`)

jason-famedly replied:
Added to docstring in [94e9f0c](https://github.com/element-hq/synapse/pull/18999/commits/94e9f0cd6cb65ee4e9bec63614935a089cccc542)

Hopefully that wasn't to much...

MadLittleMods replied:
I think it looks decent :+1:  The `INSERT` example is nice.

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -525,7 +579,7 @@ def _do_execute(
             sql_logger.debug("[SQL time] {%s} %f sec", self.name, secs)
             sql_query_timer.labels(
-                verb=sql.split()[0], **{SERVER_NAME_LABEL: self.server_name}
+                verb=one_line_sql.split()[0], **{SERVER_NAME_LABEL: self.server_name}
```
Why the change? Theoretical optimization dealing with a smaller string?

jason-famedly replied:
It may be that this was a missed hold out that could be backed out from when using the composable `SQL` class and it not being compatible here. Not certain. It likely can be reverted.

Speaking of theoretical optimization though, what do you think about adding a `maxsplit=1` to the `split()`? If memory serves that means it can bail out early and that certainly feels like it fits in the optimization category

jason-famedly replied:
swapped back to `sql` variable and put in the `maxsplit=1` micro optimization here [8584520](https://github.com/element-hq/synapse/pull/18999/commits/858452076548030f87266ac5fda054327c698049)

MadLittleMods replied:
We don't need any extra pre-optimizations added to this PR

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -1243,6 +1297,14 @@ def simple_insert_many_txn(
+                table,
+                ", ".join(k for k in keys),
+            )
+            txn.copy_write(sql, (), values)
```
Comment why

jason-famedly replied:
Added comments in [673461a](https://github.com/element-hq/synapse/pull/18999/commits/673461a388443ad354db7dfa8e51525cdc703a2f)

Some rudimentary benchmarks show that in some cases, for simple INSERT behavior like this, it may be even *faster* than `execute_values()` was

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -1243,6 +1297,14 @@ def simple_insert_many_txn(
+            )
+            txn.copy_write(sql, (), values)
+
         else:
```
Exact match SQLite if it won't work generally

jason-famedly replied:
Fortunately, it does work generally for all engines. SQLite and psycopg2 both would use a `for` loop inside which has terrible performance. 
However, psycopg v3+ can use Postgres in pipeline mode which is far better than a basic `for` loop. But it is not as performant by almost an order of magnitude as using `COPY FROM`, which is why that is preferred above(part of the incoming comment requested from [right above](https://github.com/element-hq/synapse/pull/18999/changes#r3314234658) in the full diff view)

MadLittleMods replied:
:thinking: hmmmm. We also have to think about this code evolving to adding more database drivers. Can we always make the assumption that it would work with them as well.

I feel like we should just be specific since that's our intention (only use this fallback for SQLite)

jason-famedly replied:
Then I will re-use the `assert_never()` pattern from above

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -2471,7 +2535,7 @@ def simple_delete_many_batch_txn(
         """
 
-        if isinstance(txn.database_engine, PostgresEngine):
+        if isinstance(txn.database_engine, Psycopg2Engine):
```
Why specific to `Psycopg2Engine`? `execute_values` supports both Postgres driver flavors

jason-famedly replied:
This is for `DELETE` which is not the same category as needed for `fetch=True` on `execute_values()`. Therefore, it will not function at this time. The `execute_batch()` fallback for `psycopg` will lean on `executemany()` in pipeline mode. While not as fast as a `execute_values()` it is better than a `for` loop

MadLittleMods replied:
We should add a `FIXME` for when this does work

MadLittleMods commented on pyproject.toml:
```diff
@@ -127,6 +127,11 @@ postgres = [
+# bookworm. This libpq5 includes support for postgres 18 and psycopg will fail to build
+# below this selected version.
+# See: https://github.com/psycopg/psycopg/pull/1084
+psycopg = ["psycopg[c]>=3.2.8"]
```
Comment to explain why `c` extra

Other discussion: https://github.com/element-hq/synapse/pull/18999#discussion_r2827656835

jason-famedly replied:
Tried to do this some justice in [e39c61c](https://github.com/element-hq/synapse/pull/18999/commits/e39c61c2629c288d87313c111395de3e96c3eb63)

MadLittleMods replied:
*squints* I think that's good :fast_forward:

jason-famedly committed:
Merge branch 'develop' into jason/update-psycopg3

jason-famedly committed:
expand on docstrings and add some comments

jason-famedly committed:
switch from autocommit attribute to property setter to work around __setattr__() missing

jason-famedly committed:
add assert in create_index_sqlite() to make sure it is not being erroneously used by psycopg

jason-famedly committed:
remove use of one_line_sql to split the verb out of, and add maxsplit=1 for a potential performance micro-enhancement

jason-famedly committed:
remove dead commented-out code

jason-famedly committed:
more comments

jason-famedly committed:
lint to remove unused dependency

jason-famedly committed:
add comment about why using the "c" extra and not the other implementations

jason-famedly committed:
add assert_never() statement at the end of execute_values(), to prevent accidents with database engines

jason-famedly committed:
assert that inner values are the same length

jason-famedly committed:
add assertion that template is not used yet and comment why

jason-famedly committed:
try and summarize why using COPY TO as a replacement for execute_values()

jason-famedly committed:
try and comment why simple_insert_many_txn() uses COPY FROM as it's psycopg preference

jason-famedly committed:
Merge branch 'develop' into jason/update-psycopg3

jason-famedly committed:
add enhanced docstring to new copy_write() helper on LoggingTransaction

clokep committed:
Merge pull request #10 from jason-famedly/jason/update-psycopg3

Jason/update psycopg3

jason-famedly commented on synapse/storage/database.py:
```diff
@@ -449,17 +467,120 @@ def execute_values(
+            # effort for the few places we set fetch = False.
+            assert fetch is True
+
+            # For the moment, no code paths in Synapse use `template` with `fetch=True`
+            # but the opposite is true. Prevent this mistake until the code can handle
+            # templates as well.
+            assert template is None
```
This is new. Just want to make sure reviewer doesn't miss this subtle detail

More context: https://github.com/element-hq/synapse/pull/18999#discussion_r3398968491

MadLittleMods commented on synapse/storage/engines/_base.py:
```diff
@@ -123,11 +126,14 @@ def attempt_to_set_autocommit(self, conn: ConnectionType, autocommit: bool) -> N
-        Note: This has no effect on SQLite3, as transactions are SERIALIZABLE by default.
+        Note:
+            * This has no effect on SQLite3, as transactions are SERIALIZABLE by default.
+            * On Postgres, an isolation_level of None restores the default from the
+              attribute `default_isolation_level`
```
This isn't a Postgres specific behavior. The same happens with SQLite but SQLite is just always `SERIALIZABLE`

MadLittleMods replied:
Seems like something better placed as part of the `Args:` docs

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -449,17 +467,120 @@ def execute_values(
+                    assertion_length = len(_inner_value)
+                assert assertion_length == len(_inner_value)
+
+            # To avoid having to port several psycopg2 utilities that are built into its
+            # Cursor class(mogrify, for example) and import execute_values() from it's
+            # 'extras' module, use a different mechanism that facilitates performance.
```
```suggestion
            # To avoid having to port several psycopg2 utilities that are built into its
            # Cursor class (like `mogrify`, for example) in order to re-use `from psycopg2.extras import execute_batch`,
            # use a different mechanism that is still performant.
```

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -449,17 +467,120 @@ def execute_values(
+            # The COPY Postgres-only verb allows for a bulk import and export of data.
+            # However building this query for use with VALUES is somewhat convoluted.
+            #
+            # For exporting of data, which would be the equivalent of a SELECT query,
+            # and given a simple query of the sort:
+            #  SELECT * FROM table, (VALUES ?) AS ld(id) WHERE table.id = ld.id
+            # and VALUES being an example of sequential numbers, 1-5 representing the 5
+            # rows to retrieve, the "VALUES ?" clause needs to be expanded to
+            #  VALUES (?), (?), (?), (?), (?)
+            # Then, the values themselves have to be flattened in a similar fashion
+            #  [(1, 2, 3, 4, 5)]
+            # Similarly, disregarding that the WHERE clause will no longer hold true, if
+            # the count VALUES to be passed were 3, then VALUES clause would expand to:
+            #  VALUES (?, ?, ?), (?, ?, ?), (?, ?, ?), (?, ?, ?), (?, ?, ?)
+            # and for berevity assume that all passed values were actually the same, the
+            # values would be flattened to look like:
+            #  [(1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4, 4, 5, 5, 5)]
```
Readable formatting:

```suggestion
            # For exporting of data, which would be the equivalent of a SELECT query,
            # and given a simple query of the sort:
            # ```
            # SELECT * FROM table, (VALUES ?) AS ld(id) WHERE table.id = ld.id
            # ```
            #
            # and VALUES being an example of sequential numbers, 1-5 representing the 5
            # rows to retrieve, the "VALUES ?" clause needs to be expanded to:
            # ```
            # VALUES (?), (?), (?), (?), (?)
            # ```
            #
            # Then, the values themselves have to be flattened in a similar fashion
            # ```
            # [(1, 2, 3, 4, 5)]
            # ```
            #
            # Similarly, disregarding that the WHERE clause will no longer hold true, if
            # the count VALUES to be passed were 3, then VALUES clause would expand to:
            # ```
            # VALUES (?, ?, ?), (?, ?, ?), (?, ?, ?), (?, ?, ?), (?, ?, ?)
            # ```
            #
            # and for brevity assume that all passed values were actually the same, the
            # values would be flattened to look like:
            # ```
            # [(1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4, 4, 5, 5, 5)]
            # ```
```

But overall, this is hard to understand (I haven't invested enough time to digest it) and could use some work.

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -449,17 +467,120 @@ def execute_values(
+            # Similarly, disregarding that the WHERE clause will no longer hold true, if
+            # the count VALUES to be passed were 3, then VALUES clause would expand to:
+            #  VALUES (?, ?, ?), (?, ?, ?), (?, ?, ?), (?, ?, ?), (?, ?, ?)
+            # and for berevity assume that all passed values were actually the same, the
```
```suggestion
            # and for brevity assume that all passed values were actually the same, the
```

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -385,6 +392,8 @@ def execute_batch(self, sql: str, args: Iterable[Iterable[Any]]) -> None:
                 lambda the_sql: execute_batch(self.txn, the_sql, args), sql
             )
+
+            # TODO Can psycopg3 do anything better?
```
Still relevant

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -1627,6 +1690,8 @@ def simple_upsert_many_txn_native_upsert(
 
             txn.execute_values(sql, args, fetch=False)
 
+            # TODO Maybe improve for psycopg.
```
Still relevant

MadLittleMods commented on synapse/storage/databases/main/user_directory.py:
```diff
@@ -693,37 +698,77 @@ def _update_profiles_in_user_dir_txn(
         if isinstance(self.database_engine, PostgresEngine):
```
Looks like we should just remove this extra if-layer in favor of `if isinstance(self.database_engine, Psycopg2Engine):`  and `elif isinstance(self.database_engine, PsycopgEngine):`

MadLittleMods commented on synapse/storage/databases/main/user_directory.py:
```diff
@@ -693,37 +698,77 @@ def _update_profiles_in_user_dir_txn(
+                    fetch=False,
+                )
+            elif isinstance(self.database_engine, PsycopgEngine):
+                # Unlike psycopg2, psycopg(v3+) can not use `execute_values()` at this
+                # time. However, `executemany()` on psycopg uses it's internal pipeline
+                # mode to be much faster than an iterative `execute()`, especially on
+                # writes.
```
There should be a `FIXME` comment to remove this customization once XYZ

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -424,21 +437,26 @@ def execute_batch(self, sql: str, args: Iterable[Iterable[Any]]) -> None:
             # suggests that the outer collection may be iterable, but
             # https://docs.python.org/3/library/sqlite3.html?highlight=sqlite3#how-to-use-placeholders-to-bind-values-in-sql-queries
             # suggests that the inner collection should be a sequence or dict.
+            # In the case of psycopg v3+ usage, `executemany()` uses a postgres
+            # optimization called pipelining to vastly speed up processing of the query
+            # when there are many args.
```
```suggestion
            #
            # In the case of psycopg v3+ usage, `executemany()` uses a postgres
            # optimization called pipelining to vastly speed up processing of the query
            # when there are many args.
```

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -499,6 +499,9 @@ def f(
             # Flatten the values.
             return self._do_execute(f, sql, list(itertools.chain.from_iterable(values)))
 
+        # mypy does not like that self.database_engine is not a Never type
+        assert_never(self.database_engine)  # type: ignore[arg-type]
```
🤔 I think the type ignore is the wrong thing to do here. The whole point of `assert_never(...)` is that it will catch it at the lint stage as well.

Although I see that `assert_never` does something at runtime as well so maybe this is fine (would be just as equivalent to the `raise AssertionError(...)` we would replace it with anyway.

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -449,17 +467,120 @@ def execute_values(
+
+            # execute_values requires a single replacement, but we need to expand it
+            # for COPY. These inner sequences must be the same length.
+            assertion_length = 0
```
```suggestion
            assertion_length: int | None  = None
```

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -449,17 +467,120 @@ def execute_values(
+                # types wanted while excluding Generators and this assertion can be
+                # removed.
+                assert isinstance(_inner_value, Sized)
+                if not assertion_length:
```
```suggestion
                if assertion_length is None:
```

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -449,17 +467,120 @@ def execute_values(
+                assert isinstance(_inner_value, Sized)
+                if not assertion_length:
+                    assertion_length = len(_inner_value)
+                assert assertion_length == len(_inner_value)
```
We can add an assertion message explaining what's wrong and what to do

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -449,17 +467,120 @@ def execute_values(
+        self, sql: str, args: Iterable[Any], values: Iterable[Iterable[Any]]
+    ) -> None:
+        """
+        Corresponds to a PostgreSQL COPY (...) FROM STDIN call using psycopg v3
```
```suggestion
        Corresponds to a PostgreSQL `COPY (...) FROM STDIN` call using psycopg v3
```

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -449,17 +467,120 @@ def execute_values(
+        it would be
+            COPY table (user_id, email, locked) FROM STDIN
+
+        The 'values' parameter input should not change, a Collection of Iterables.
+        Details on formatting are handled by psycopg.
```
What is this trying to say?

Seems like this should be a `Sequence` instead of an `Iterable` if we're trying to say we should be able to iterate over it multiple times (perhaps probably because of retries?)

jason-famedly replied:
Well it has been a little while, so I'm not completely certain now either. I think I was trying to convey that nothing special needs to be done for `values`, such as pre-flattening or wrapping in an extra `Iterable`. Type narrowing was to be done in a separate work, but ultimately would like this to be a minimum of a `Collection`. The inner workings of how `values` interact inside of the transaction retry semantics of Synapse don't tell me enough to know if the `values` needs to have an index(`__getitem__()`) or not. 

Not sure where you would otherwise get that it would iterated over multiple times...?

jason-famedly replied:
I will happily take suggestions on which way you want this to read. Type narrowing is later, since this is still an experimental phase. Apparently the precursor to this patch where some type narrowing was done did apply `Sequence`, so I am definitely not against that if you want the comment to reflect such

MadLittleMods commented on pyproject.toml:
```diff
@@ -127,6 +127,18 @@ postgres = [
     "psycopg2cffi-compat==1.1;platform_python_implementation == 'PyPy'",
 ]
+# The "pure python" implementation of psycopg(v3+) does not play nicely with the
+# complement docker image because of differences in the required debian package `libpq`.
```
```suggestion
# Complement Docker image because of differences in the required Debian package `libpq`.
```

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -449,17 +467,120 @@ def execute_values(
+                values,
+            )
+        elif isinstance(self.database_engine, PsycopgEngine):
+            # We use fetch = False to mean a writable query. You *might* be able
+            # to morph that into a COPY (...) FROM STDIN, but it isn't worth the
+            # effort for the few places we set fetch = False.
+            assert fetch is True
```
```suggestion
            # We use `fetch=False` to mean a writable query. You *might* be able
            # to morph that into a `COPY (...) FROM STDIN`, but it isn't worth the
            # effort for the few places we set `fetch=False`.
            assert fetch is True
```

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -449,17 +467,120 @@ def execute_values(
+                values,
+            )
+        elif isinstance(self.database_engine, PsycopgEngine):
+            # We use fetch = False to mean a writable query. You *might* be able
+            # to morph that into a COPY (...) FROM STDIN, but it isn't worth the
+            # effort for the few places we set fetch = False.
+            assert fetch is True
```
LLM spotted that this crashes with the `threads_backfill_txn`

https://github.com/element-hq/synapse/blob/3f3edaf0293c9b85ce25451daeb9acd926820a3a/synapse/storage/databases/main/relations.py#L141-L142

jason-famedly replied:
It could probably, yes. The [change that introduced that](https://github.com/matrix-org/synapse/pull/13394) was for populating a new table at the time, `threads`, which is now filled proactively instead of being backfilled from `event_relations`. That was about Oct, 2022. Should this be removed or supported for longer?

MadLittleMods commented on synapse/_scripts/synapse_port_db.py:
```diff
@@ -1556,7 +1556,7 @@ def main() -> None:
         sys.exit(2)
-    if postgres_config["name"] != "psycopg2":
+    if postgres_config["name"] not in ("psycopg", "psycopg2"):
         sys.stderr.write("Database must use the 'psycopg2' connector.\n")
```
```suggestion
        sys.stderr.write("Database must use the 'psycopg' or 'psycopg2' connector.\n")
```

MadLittleMods commented on synapse/_scripts/synapse_port_db.py:
These kind of queries supposedly fail with `psycopg` (previously compatible with `psycopg2`).

See https://www.psycopg.org/psycopg3/docs/basic/from_pg2.html#server-side-binding

Problem spots:

 - A few examples here in `synapse/_scripts/synapse_port_db.py`
 - and `synapse/storage/schema/main/delta/58/11user_id_seq.py`

jason-famedly replied:
The first example, yes I will give you. Working on the db porting script was slated for later, but I will go ahead and do that now.

The second: this is true, and this will have to be sorted out when the next time it comes that it is an issue. The current full schema version that is applied before deltas is 72(generated Sept of 2022). The youngest `CREATE SEQUENCE` *that is ran in python without a hardcoded script* is at 69(this is an important detail because of parameters). The rest of the newer are part of the raw files that are ran through the `executescript()` function and the values are hardcoded and not parameterized, so this is not an issue there.

MadLittleMods commented on synapse/storage/engines/psycopg.py:
```diff
@@ -0,0 +1,96 @@
+        # Disables passing `bytes` to txn.execute, c.f. #6186. If you do
+        # actually want to use bytes than wrap it in `bytearray`.
+        # def _disable_bytes_adapter(_: bytes) -> NoReturn:
+        #     raise Exception("Passing bytes to DB is disabled.")
```
I think this kind of check is important to ensure we're not accidentally doing the wrong thing.

In other words, we don't want bytes that were accidentally passed to us to be silently written as corrupted hex-blob text in the database.

Does `psycopg` handle this for us? Test?

MadLittleMods commented on changelog.d/18999.feature:
```diff
@@ -0,0 +1 @@
+Support using psycopg as the PostgreSQL driver. Contributed by @clokep and @realtyem.
```
LLM spotted that we're using a `psycopg2` specific error here:

https://github.com/element-hq/synapse/blob/3f3edaf0293c9b85ce25451daeb9acd926820a3a/synapse/storage/databases/main/search.py#L286-L302

jason-famedly replied:
Looks like it was correct. That section is also over 8 years old now, and `CREATE INDEX CONCURRENTLY IF NOT EXISTS` has been supported since Postgres 9.5(2016 or so). How much longer should this index change be supported?

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -424,21 +437,26 @@ def execute_batch(self, sql: str, args: Iterable[Iterable[Any]]) -> None:
+            # when there are many args.
             self.executemany(sql, args)
 
     def execute_values(
```
Does this LLM review point make sense? It seems plausible but I haven't fully onboarded to the details yet:

 - `synapse/storage/database.py` (line 547)
 - **Summary:** "The PsycopgEngine execute_values branch returns a lazy generator from _do_execute — the COPY does not execute inside _do_execute's timing/tracing/error handling, and does not execute at all unless the caller iterates the result (psycopg2's execute_values executed eagerly and returned a list)",
 - **Failure scenario:** "Under psycopg, sql_query_timer records ~0s and the opentracing db.query span closes before the query runs; DB errors (e.g. serialization_failure in the auth-chain query at event_federation.py:350) surface at the consumer's iteration site bypassing [SQL FAIL] logging, and any future caller that discards the return value gets a query that silently never executes"

clokep replied:
That sounds plausible, we could easily do a `return list[...]` for the results, but that kind of defeats the entire performance optimization.

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -424,21 +437,26 @@ def execute_batch(self, sql: str, args: Iterable[Iterable[Any]]) -> None:
+            # when there are many args.
             self.executemany(sql, args)
 
     def execute_values(
```
Does this LLM review point make sense? It seems plausible but I haven't fully onboarded to the details yet:

 - `synapse/storage/database.py` (line 535)
 - **Summary:** "The psycopg execute_values rewrites the query with sql.replace('?', ...), expanding EVERY '?' rather than the single 'VALUES ?' slot, returns all columns as unparsed strings (copy.rows() without set_types — empirically 42 → '42'), and inlines all rows into one unbounded statement where psycopg2 paginated at 100 rows",
 - **Failure scenario:** "A future caller whose query has a second '?' (e.g. an extra WHERE param) gets it silently expanded into the VALUES list with misaligned bindings; any caller selecting a non-text column gets str instead of int/bool; and today's auth-chain path (event_federation.py:901) can build a multi-megabyte single COPY statement for tens of thousands of chains where psycopg2 executed bounded 100-row pages"

clokep replied:
I'm not really sure what this is saying. I think the columns will still be returned of the correct types, but I'm not 100% sure on that (it has been way too long since I wrote this code).

MadLittleMods commented on synapse/storage/engines/psycopg.py:
```diff
@@ -0,0 +1,96 @@
+            return error.sqlstate in ["40001", "40P01"]
+        return False
+
+    def in_transaction(self, conn: psycopg.Connection) -> bool:
+        return conn.info.transaction_status != psycopg.pq.TransactionStatus.IDLE
```
This is something to double-check as it's different than pscyopg2:

https://github.com/element-hq/synapse/blob/3f3edaf0293c9b85ce25451daeb9acd926820a3a/synapse/storage/engines/postgres.py#L230-L231

Possible values: [`psycopg/psycopg` -> `psycopg/psycopg/pq/_enums.py#L133-L153`](https://github.com/psycopg/psycopg/blob/6a52e77dc38827214b294759823a072841b4819f/psycopg/psycopg/pq/_enums.py#L133-L153)
```python
class TransactionStatus(IntEnum):
    """
    The transaction status of a connection.
    """

    __module__ = "psycopg.pq"

    IDLE = 0
    """Connection ready, no transaction active."""

    ACTIVE = auto()
    """A command is in progress."""

    INTRANS = auto()
    """Connection idle in an open transaction."""

    INERROR = auto()
    """An error happened in the current transaction."""

    UNKNOWN = auto()
    """Unknown connection state, broken connection."""
```

Can we use `psycopg.pq.TransactionStatus.INTRANS` instead?

The current logic at-least looks flawed as it will consider `psycopg.pq.TransactionStatus.UNKNOWN` as in a transaction.

jason-famedly replied:
Based on [the `psycopg2` docs](https://www.psycopg.org/docs/extensions.html#transaction-status-constants) which appear to mirror what is found in what you linked, I find nothing to do differently here.

Additionally, this is used in exactly one spot => starting a transaction. Even if it is an unknown state, that should not proceed
